rfd/0004-workspace-and-prebuild-storage.md

```diff
@@ -0,0 +1,97 @@
+  - code images (based on prebuild code images)
+  - os images (based on prebuild os images)
+
+In total, there are 5 types of images Hocus manages.
+
+### Local and remote storage of images
+
+Prebuild images are pushed to a user-configurable OCI registry after they are built. Every prebuild has a tag that is partly deterministic, which allows us to check if it's been built before and if the agent can download it instead of building. After they are downloaded or built, prebuild images are stored on disk until the associated prebuild is archived. Workspace images are based on prebuild images, but for now they are only stored locally and never pushed to a remote location. They are deleted once a workspace is deleted. Git repository images are not pushed to a remote registry for now.
+
+Hocus does not remove tags from the remote registry. It's up to the user to configure a retention period in the registry themselves.
+
+### Prebuild image tags
+
+Prebuilds consist of two images: one that contains the git repository, and one that contains all the other OS files. Prebuild image tags have the following format:
+
+```js
+`${deterministicPart}-prebuild-${"repo" | "os"}-${randomId}`;
+```
+
+The deterministic part is the result of:
+
+```js
+sha256(`${gitRepositoryUrl}-${gitObjectHash}-${projectRootDir}`);
+```
+
+The deterministic part lets us check if a prebuild was completed before and if the agent can download it instead of building. The agent does it by computing the deterministic part and then querying the [`/v2/<name>/tags/list?n=<int>&last=<deterministicPart>-` registry endpoint](https://github.com/opencontainers/distribution-spec/blob/8e2a64e837af0392c0eabe7f36f79ff621159fe7/spec.md#listing-tags). For example, here's how to do it with curl:
+
+```bash
+response=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:hocusdev/workspace:pull")
+token=$(echo "$response" | jq -r '.token')
+export DOCKER_REGISTRY_TOKEN="$token"
+curl -H "Authorization: Bearer $DOCKER_REGISTRY_TOKEN" "https://registry-1.docker.io/v2/hocusdev/workspace/tags/list?n=5&last=02-06-2023"
+```
+
+The endpoint returns a list of `n` tags in lexicographical after the `last` tag, but not including it. If any tags with the deterministic part prefix are returned, the agent can download one of them instead of building.
+
+### Workspace image tags
+
+Workspaces are created from prebuilds, and they also consist of two images. Workspace image tags have the following format:
+
+```js
+`${deterministicPart}-workspace-${"repo" | "os"}-${randomId}`;
+```
+
+The deterministic part is the same as the underlying prebuild's.
+
+### Git repository image tags
+
+Their tag format is:
+
+```js
+${deterministicPart}-${lastFetchTimestampSeconds}-${randomId}
+```
+
+It's likely to change in the future when we start coordinating agents to reuse git repository images more efficiently.
+
+The deterministic part is the result of
+
+```js
+sha256(gitRepositoryUrl);
```

Comment:
Why not reuse gitObjectHash?

Reply:
Right now this tag could be an UUID, so it doesn't really matter.